src/builder/specification.md

```diff
@@ -0,0 +1,316 @@
+* Builder API messages, e.g. [`builder_registerValidatorV1`](#builder_registerValidatorV1) and the response to [`builder_getHeader`](#response-2), which should compute the signing root using [`compute_signing_root`][compute-signing-root] and the domain `DomainType('0xXXXXXXXX')` (TODO: get a proper domain).
+
+As `compute_signing_root` takes `SSZObject` as input, client software should convert in-protocol messages to their SSZ representation to compute the signing root and Builder API messages to the SSZ representations defined [above](#sszobjects).
+
+## Methods
+
+### `builder_status`
+
+#### Request
+
+- method: `builder_status`
+- params: `null`
+
+#### Response
+
+- result: `enum`, `"OK" | null`
+- error: code and message set in case the builder is not operating normally.
+
+#### Specification
+
+1. Builder software **MUST** return `-32000: Server error` if it is unable to respond to requests. `err` **SHOULD** be set explaining the issue and `result` **MUST** be `null`.
+
+### `builder_registerValidatorV1`
+
+#### Request
+
+- method: `builder_registerValidatorV1`
+- params:
+  1. `message`: `object`
+      1. `feeRecipient`: `DATA`, 20 Bytes - Address of account which should receive fees.
+      2. `gasLimit`: `QUANTITY`, 64 bits - Target gas limit for block.
+      3. `timestamp`: `QUANTITY`, 64 bits - Unix timestamp of announcement.
+      4. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
+  2. `signature`: `DATA`, 96 Bytes - BLS signature over [`ValidatorRegistrationV1`](#validatorregistrationv1).
+
+#### Response
+
+- result: `enum`, `"OK" | null`
+- error: code and message set in case an exception happens while registering the validator.
+
+#### Specification
+1. Builder software **MUST** verify `pubkey` corresponds to an active or pending validator, otherwise return error `-32002: Unknown validator`.
+2. Builder software **MUST** verify `signature` is valid under `pubkey`, otherwise return error `-32005: Invalid Signature`.
+3. Builder software **MUST** respond to requests where `timestamp` is less than or equal to the latest announcement from the validator or more than 10 seconds in the future with error `-32007: Invalid timestamp`.
+4. Builder software **MUST** return `result` as `"OK"` if the request succeeds, `null` otherwise.
+
+### `builder_getHeaderV1`
+
+#### Request
+
+- method: `builder_getHeaderV1`
+- params:
+  1. `slot`: `QUANTITY`, 64 Bits - Slot number of the block proposal.
+  2. `pubkey`: `DATA`, 48 Bytes - BLS public key of validator.
+  3. `parentHash`: `DATA`, 32 Bytes - Hash of execution layer block the proposer will use as the proposal's parent.
+
+#### Response
+
+- result: `object`
+    - `message`: `object`
+        - `header`: [`ExecutionPayloadHeaderV1`](#executionpayloadheaderv1).
+        - `value`: `QUANTITY`, 256 Bits - Payment in wei that will be paid to the `feeRecipient` account.
+        - `pubkey`: `DATA`, 48 Bytes - BLS public key associated with the builder.
```

Comment:
Sidenote: this might actually be the pubkey of the relay, or we might want signatures by both builder and relay. Definitely need the relay signature if the validator wants to whitelist a specific relay and require correct signatures from the relay.